doc/developer/design/20210707_qgm_sql_high_level_representation.md

```diff
@@ -0,0 +1,702 @@
+queries as their input and a list of transformations to be applied, and will dump the content of the graph after
+applying each of them. The query graph will be dumped as a `dot` graph which, once rendered, looks like the
+ones in the appendix with examples.
+
+The `dot` graph generator will also be very handy for debugging transformations. Also, for better debuggability
+Jthe rules themselves should not drive the traversal of the graph, but just make explicit the type of traversal
+they require. The traversal of the graph will be driven by some external code, that could be instrumentalized for
+dumping the full graph before and after applying the rule to every node of the graph. That way, by just patching
+one method, we could follow step by step the evolution of a query graph as it goes through the different
+transformations.
+
+### Lowering
+
+The lowering code will have to be modified so that it takes a Query Graph as its input.
+
+When lowering a `Select` box we will first build the join within that box. To do so, we will first build a
+dependency graph where each quantifier is connected to the quantifiers it is correlated with. We will lower
+the quantifiers in dependency order, starting with the uncorrelated ones. When lowering
+a correlated quantifier, we will apply its plan to the result of the sub-join of all the quantifiers it is
+connected to, resulting in simpler dataflow plans for lateral joins.
+
+The dependency graph will be built by traversing the sub-graph of the input quantifiers collecting the
+column references from the current context.
+
+Note that the lowering code will no longer produce binary joins exclusively.
+
+After lowering the join, a `Map` operator will be added with the non-column expressions in the projection
+of the box. Then, a `Projection` will be added. If the box must enforce distinctness, the corresponding
+`Reduce` operator will be added. If the box has `LIMIT`/`OFFSET`/`ORDER BY` a `TopK` operator will
+then be added.
+
+### View inlining
+
+View inlining could be implemented as a normalization rule where only non-materialized views are inlined.
+
+We could also do the opposite: find query patterns that match existing materialized views. Furthermore,
+we could even have a view compaction step where we load all the views in the system into the same model
+and find common sub-expressions within it, to identify intermediate results worth materializing.
+
+## Alternatives
+
+<!--
+// Similar to the Description section. List of alternative approaches considered, pros/cons or why they were not chosen
+-->
+
+* QGM with interior mutability, shared pointers and so on as implemented [here](https://github.com/asenac/rust-sql).
+    * Making all boxes to be directly owned by the model makes recursion easier to support.
+* Relational algebra representation
+* Convert `MirRelationExpr` into a normalization-friendly representation with explicit `outer join` operator
+  and fewer number of operators.
+
```

Comment:
So the existing `Scope` stuff has all sorts of horrible hacks that I jammed in over time. For example, this `priority` bool:

https://github.com/MaterializeInc/materialize/blob/dd2e68c908d42924a5cf5a439dbfcfcab3fb27da/src/sql/src/plan/scope.rs#L57-L80

And the fact that the first name of a column in the `names` vec is special: https://github.com/MaterializeInc/materialize/blob/dd2e68c908d42924a5cf5a439dbfcfcab3fb27da/src/sql/src/plan/scope.rs#L85-L89

And the fact that we have columns that have names but are not nameable: https://github.com/MaterializeInc/materialize/blob/dd2e68c908d42924a5cf5a439dbfcfcab3fb27da/src/sql/src/plan/scope.rs#L91-L96

I don't understand these things deeply enough to tell you what the key insight is of `NameResolutionContext` that will make it avoid these things (or if it might in fact end up with some hacks of its own), but there is basically no universe in which it is worse :).

Reply:
Oh, I wasn't aware of this subtle difference around natural joins that makes the common column `f1` non-ambiguous:

```
asenac=> select * from t1 natural left join t2;
 f1 | f2 
----+----
  1 |  2
  2 |  2
(2 rows)

asenac=> select * from t1 left join t2 on t1.f1 = t2.f1;
 f1 | f2 | f1 
----+----+----
  1 |  2 |  1
  2 |  2 |   
(2 rows)

asenac=> select f1 from t1 natural left join t2;
 f1 
----
  1
  2
(2 rows)

asenac=> 
```

Yeah, that would require a hack of its own :(

Reply:
This commit against the prototyping branch contains how I'd handle the intrusive ORDER BY clauses:

https://github.com/asenac/materialize/commit/3ebf5b354df281ee975723d7e2d525e081f0940f

A `use_box_projection` flag is added to `NameResolutionContext` that, when enabled, uses the projection of the current `Select` for name resolution and check for ambiguities. When testing it, I found #7549.